packages/playwright-core/src/utils/tar.ts

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import fs from 'fs';
+import { Writable, once } from 'stream';
+import assert from 'assert';
+
+enum TarType {
+  REGTYPE,
+  LNKTYPE,
+  SYMTYPE,
+  CHRTYPE,
+  BLKTYPE,
+  DIRTYPE,
+  FIFOTYPE,
+  CONTTYPE
+}
+
+class TarEntry {
+  name: string;
+  size: number;
+  type: TarType;
+  mode: number;
+  linkname: string;
+  uid: number;
+  gid: number;
+
+  fileStream: fs.WriteStream | null = null;
+  remainingBytes = 0;
+
+  constructor(header: Buffer) {
+    if (header.length < 512)
+      throw new Error('Invalid header: ' + header.toString('utf8'));
+
+    this.name = header.toString('utf8', 0, 100).replace(/\0/g, '').replace(/^\/+/, '');
+
+    this.size = parseInt(header.toString('ascii', 124, 136).trim(), 8);
+    this.type = parseInt(header.toString('ascii', 156, 157), 10) as TarType;
+    this.mode = parseInt(header.toString('ascii', 100, 108).trim(), 8) || 0o644;
+    this.linkname = header.toString('utf8', 157, 257).replace(/\0/g, '');
+
+    this.uid = parseInt(header.toString('ascii', 108, 116).trim(), 8);
+    this.gid = parseInt(header.toString('ascii', 116, 124).trim(), 8);
+  }
+
+  async writeToDisk(outputPath: (path: string) => string) {
+    const fullPath = outputPath(this.name);
```

Comment:
why is `outputPath` a callback? Ideally its a string and we should add a check so that users can't write 'outside' of the outputPath.

Reply:
I made this a callback instead of a raw string, so that the calling side can implement logic like this:

https://gist.github.com/mxschmitt/0807f729ba68358e58bf5210b4607fdc#file-har-extract-mjs-L71-L76

It seemed important that we can do that, but at the same time it also didn't seem like a property of the TAR extraction, so it can live outside of it now.